### PR TITLE
Feature/storage/sas client derived

### LINKED ISSUE #29742: [BUG] GetBlobClient or any GetStorageClient causes SAS to be parsed and reordered
### Library name and version

Azure.Storage.Blobs; Azure.Storage.Queues; Azure.Storage.Files.Shares; Azure.Storage.Files.DataLake

### Describe the bug

When calling .GetBlobClient(blobname) when the Parent client is authenticated with a SAS token in the URL, the SAS will be manipulated and reformed

We need to check all GetStorageClients to check if the SAS is getting parsed and put back together. What needs to be done is saving the original sas token when in the URI builder instead of saving it as a BlobSasQueryParameters.

Note: This does not apply to GetContainerClient because it does not call BlobUriBuilder.

Mitigation 1 - Use AzureSasCredential
```
BlobServiceClient serviceClient = new BlobServiceClient(new Uri("https://storageaccount.blob.storage.azure.net"), new AzureSasCredential(sas), options);
BlobContainerClient containerClient = serviceClient.GetContainerClient(containername);
BlobClient blobClient = containerClient.getBlobClient(blobname);
```

Mitigation 2 - Declare Client without using GetStorageClient
```
BlobClient serviceClient = new BlobClient ("[https://storageaccount.blob.storage.azure.net/?<sas token>", options);
```

### Expected behavior

Do not manipulate SAS, send the SAS as is and also do not manipulate SAS between clients.

### Actual behavior

The SAS gets manipulated between clients.

### Reproduction Steps

BlobServiceClient serviceClient = new BlobServiceClient(new Uri(https://storageaccount.blob.storage.azure.net/?<sas token>srt="cos"));

BlobClient blobClient = serviceClient.GetBlobContainerClient(containername).GetBlobClient(blobname);

//(Calling only GetBlobContainerClient will not reproduce the bug because BlobUriBuilder is not getting called underneath to reform the URL)

### Environment

n/a

### PR DESCRIPTION
**Pull request by amnguye**
Resolves https://github.com/Azure/azure-sdk-for-net/issues/29742

Also there was a bug in ShareClient and DataLakeFileSystemClient's API GetParentServiceClient would return an service client with a URL that persisted the container name. I resolved those as well and updated the changelogs.

### CODE REVIEW

**Comment by amnguye** [comment]
TODO: reduce unnecessary tests that are redundant

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs [hunk 1]
```diff
@@ -572,5 +573,36 @@ public void Handle(BlobQueryError blobQueryError)
...
+        /// <summary>
+        /// Gets a custom account SAS where the permissions, services and resourceType
+        /// comes back in the string character order that the user inputs it as.
+        /// </summary>
+        /// <param name="permissions"></param>
+        /// <param name="services"></param>
+        /// <param name="resourceType"></param>
+        /// <param name="sharedKeyCredential"></param>
+        /// <returns></returns>
+        public string GetCustomAccountSas(
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs:586]: I can update the comment to something like add
```
        /// e.g. If someone has the services (srt) value as "ocs", then in the
        /// sas token returned by this method will have the srt=osc and the
        /// string-to-sign will be signed with the services value as osc.
        ///
        /// This is custom because using the regular AccountSasBuilder does not accept
        /// a raw string services or resourceTypes, it accepts services and resourceTypes
        /// (e.g. see AccountSasPermissions, AccountSasServices, AccountSasResourceType).
        /// The order is fixed respective to what values is set.
```

File: sdk/storage/Azure.Storage.Common/CHANGELOG.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # Release History
 
 ## 12.13.0-beta.1 (Unreleased)
-- This release contains bug fixes to improve quality.
+- Fixed bug where Account SAS with the resources type in any order but "sco" would get reordered to "sco", which would invalidate the Account SAS signature from the string to sign
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Common/CHANGELOG.md:4]: Yup my grammar mistake, I meant to say "with any resources type in any order BUT "sco"**,** would get reordered to "sco".

File: sdk/storage/Azure.Storage.Files.DataLake/tests/DataLakeTestBase.cs [hunk 3]
```diff
@@ -379,6 +380,37 @@ public async Task<string> SetupFileSystemLeaseCondition(DataLakeFileSystemClient
...
+        /// <summary>
+        /// Gets a custom account SAS where the permissions, services and resourceType
+        /// comes back in the string character order that the user inputs it as.
+        /// </summary>
+        /// <param name="permissions"></param>
+        /// <param name="services"></param>
+        /// <param name="resourceType"></param>
+        /// <param name="sharedKeyCredential"></param>
+        /// <returns></returns>
+        public string GetCustomAccountSas(
```
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Files.DataLake/tests/DataLakeTestBase.cs:392]: I threw this method in every service test base to prevent having to include all the necessary shared SAS test resources, that took a dependency on some SAS storage shared resources. Let me try again to throw this method into shared test code and remember why it was painful to do, or if I was just having a bad day with dependencies.

Also FWIW So for DataLake specifically, it doesn't allow you to set the immutable policy ('i) in the permissions. So if you attempted to send it with an 'i' in the permissions the service throws an error. So the default permissions does not include the 'i' in comparison to the rest of the services. I mean what we can do is just not include 'i' in the permissions for all of them because I doubt any of the tests require the 'i' by default.

**Comment by amnguye** [comment]
TODO: remove permissions test, because I overlooked that we publicly allow setting the raw string permissions.

File: sdk/storage/Azure.Storage.Common/CHANGELOG.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Common/CHANGELOG.md:4]: I'm just going to reword it, that way it's less confusing.

"- Fixed bug where Account SAS with the resources type value not in the order "sco" would get reordered to that order, which would invalidate the Account SAS signature from the string to sign"

File: sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs]: So as per the other comment, I moved this method to sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.SasVersion.cs
Updated the comment there.

File: sdk/storage/Azure.Storage.Files.DataLake/tests/DataLakeTestBase.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (amnguye) [reply sdk/storage/Azure.Storage.Files.DataLake/tests/DataLakeTestBase.cs]: Okay so this still kinda exists in DataLakeTestBase BUT because we need to take in the HNS credentials instead of the Primary credentials

**Review by jaschrep-msft — APPROVED** [review]
